workspaces/diff-engine/src/commands/rfc.rs

```diff
@@ -0,0 +1,262 @@
+        ))]
+      }
+
+      RfcCommand::EndBatchCommit(command) => {
+        validation.require(
+          validation.batch_commit_id_exists(&command.batch_id),
+          "batch commit must exist to end batch commit",
+        )?;
+        validation.require(
+          !validation.batch_commit_completed(&command.batch_id),
+          "batch commit can not have ended yet to end batch commit",
+        )?;
+
+        vec![RfcEvent::from(rfc_events::BatchCommitEnded::from(command))]
+      }
+
+      _ => Err(SpecCommandError::Unimplemented(
+        "rfc command not implemented for rfc projection",
+        SpecCommand::RfcCommand(self),
+      ))?,
+    };
+
+    Ok(events)
+  }
+}
+
+struct CommandValidationQueries<'a> {
+  command_description: String,
+  history_projection: &'a HistoryProjection,
+}
+
+impl<'a> CommandValidationQueries<'a> {
+  fn require(&self, condition: bool, msg: &'static str) -> Result<(), SpecCommandError> {
+    if condition {
+      Ok(())
+    } else {
+      Err(SpecCommandError::Validation(format!(
+        "Command failed validation: {}, {:?}",
+        msg, self.command_description
+      )))
+    }
+  }
+
+  pub fn batch_commit_id_exists(&self, commit_id: &CommitId) -> bool {
+    self
+      .history_projection
+      .get_batch_commit_node_index(commit_id)
+      .is_some()
+  }
+
+  pub fn batch_commit_completed(&self, commit_id: &CommitId) -> bool {
+    let descriptor = self
+      .history_projection
+      .get_batch_commit_descriptor(commit_id);
+    descriptor.is_some() && descriptor.unwrap().is_complete
+  }
+}
+
+impl<'a> From<(&'a HistoryProjection, &RfcCommand)> for CommandValidationQueries<'a> {
+  fn from((history_projection, rfc_command): (&'a HistoryProjection, &RfcCommand)) -> Self {
+    Self {
+      command_description: format!("{:?}", rfc_command),
+      history_projection,
+    }
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use super::*;
+  use crate::events::SpecEvent;
+  use cqrs_core::Aggregate;
+  use insta::assert_debug_snapshot;
+  use serde_json::json;
+
+  #[test]
+  pub fn can_handle_append_batch_command() {
+    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
+      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
+      {"BatchCommitStarted": {"batchId": "batch_1","parentId": "root", "commitMessage": "initial commit"}},
+      {"BatchCommitEnded": {"batchId": "batch_1" }}
+    ]))
+    .expect("initial events should be valid spec events");
+
+    let mut projection = HistoryProjection::from(initial_events);
+
+    let command: RfcCommand = RfcCommand::append_batch_commit(String::from("second commit"));
+
+    let new_events = projection
+      .execute(command)
+      .expect("new command should be applicable to initial projection");
+    assert_eq!(new_events.len(), 1);
+    // TODO: snapshot the results instead, once we figure out how insta redactions can control
+    // for the automatically generated batch id.
```

Comment:
i'd follow the pattern we used in scala, injecting an idgenerator. for testing, use sequential ids; otherwise use uuidv4 or something else appropriate for the likelihood of collisions

Reply:
It's what I tried initially, but the Rust ownership model makes that pretty clunky. It's exactly the kind of thing Insta's redactions feature is for, I just couldn't figure out how its redaction paths work. Given the amount of code here I didn't feel like going any deeper into that instantly, but seems like the most straightforward way of dealing with this.

Reply:
Having all of these transformations be deterministic during testing is worth it. I'm hesitant to invest in redactions because it adds friction to the testing process. snapshots should be a low friction thing and it's important to be able to reason about the output more clearly and there are a lot of ids you wouldn't want to all end up as `[uuid]`. I'm not clear on what is clunky about the id generator approach. As far as I can tell, it would be another member of the command handling aggregate's struct (I also note you didn't wrap the projections in a more semantic struct yet) and could be concisely represented as an iterator of strings.

Reply:
We're going to follow this up in a future PR. There's some friction from redactions in snapshots, but possibly also in the refactored version. We'll have to try and figure out what's best, but we'll leave that out of the scope of this first base one.